File: service-jobs/src/controllers/RabbitMq.ts
import * as ampq from 'amqplib';
import logger, { getLoggerMetaFactory } from '../logger';
import config from '../configs/server';

const makeLoggerId = getLoggerMetaFactory('RabbitMQConnection');
const connInfo = { ...config.rabbitMq.connectSettings };
const originQueueName = config.rabbitMq.serviceQueueName;

export type RabbitMqMessage<T> = {
    from: string;
    to: string;
    time: string;
    jobId: string;
    message: T;
};

export class RabbitMqConnection {
    connection?: ampq.ChannelModel;
    channel?: ampq.Channel;
    queueName: string;
    durable: boolean;

    constructor(queueName: string, durable: boolean = true) {
        this.queueName = queueName;
        this.durable = durable;
    }

    async connect() {
        const logId = makeLoggerId('connect');
        if (this.connection && this.channel) return;
        logger.info(
            `connecting to RabbitMQ Channel: ${connInfo.hostname}: ${connInfo.port} for queue: ${this.queueName}`,
            logId,
        );
        this.connection = await ampq.connect(connInfo);
        this.channel = await this.connection.createChannel();
        await this.channel.assertQueue(this.queueName, { durable: this.durable });
        logger.info('RabbitMQ Connection Established', logId);
    }

    isConnected(): boolean {
        return !!this.connection && !!this.channel;
    }

    async close() {
        logger.info(
            `closing connection to RabbitMQ on ${connInfo.hostname}:${connInfo.port} with queue: ${this.queueName}`,
        );
        await this.channel?.close();
        await this.connection?.close();
        this.channel = undefined;
        this.connection = undefined;
    }
}

export class RabbitMqConnectionManager {
    protected connection: RabbitMqConnection;
    protected queueName: string;

    constructor(queueName: string, durable: boolean = true) {
        this.queueName = queueName;
        this.connection = new RabbitMqConnection(queueName, durable);
    }

    getQueueName() {
        return this.queueName;
    }

    isConnected() {
        return this.connection?.isConnected() ?? false;
    }

    async close() {
        await this.connection?.close();
    }
}

export class RabbitMqMessageSender extends RabbitMqConnectionManager {
    async sendJsonMessage<T>(
        queueName: string,
        message: T,
        jobId: string = '',
        corrId: string = '',
        jweToken: string = '',
        persistent: boolean = true,
    ) {
        if (!this.connection.isConnected()) {
            await this.connection.connect();
        }
        const messageToSend: RabbitMqMessage<T> = {
            from: originQueueName,
            to: queueName,
            jobId,
            time: new Date().toISOString(),
            message,
        };
        const jsonBuffer = Buffer.from(JSON.stringify(messageToSend));
        this.connection.channel!.sendToQueue(queueName, jsonBuffer, {
            persistent,
            headers: {
                'x-correlation-id': corrId,
                authorization: `Bearer ${jweToken}`,
            },
        });
    }
}

export class RabbitMqMessageReceiver extends RabbitMqConnectionManager {
    private autoAcknowledge: boolean;

    constructor(queueName: string, durable: boolean = true, autoAcknowledge: boolean = false) {
        super(queueName, durable);
        this.autoAcknowledge = autoAcknowledge;
    }

    async getMessagesOnQueue<T>(
        callback: (message: RabbitMqMessage<T>, messageSource: ampq.ConsumeMessage) => void,
    ) {
        if (!this.connection.isConnected()) {
            await this.connection.connect();
        }
        const logId = {
            id: 'RabbitMQMessageReceiver/getMessagesOnQueue',
        };
        this.connection.channel!.consume(this.queueName, (message) => {
            if (message === null) return;
            logger.info(`message received on queue: ${this.queueName}`, logId);
            const parsedMessage: RabbitMqMessage<T> = JSON.parse(message.content.toString());
            callback(parsedMessage, message);
            if (this.autoAcknowledge && this.connection.isConnected()) {
                logger.debug(
                    `auto acknowledging message ${parsedMessage.to} / ${parsedMessage.from}`,
                    logId,
                );
                this.connection.channel!.ack(message);
            }
        });
    }

    async acknowledgeMessageReceipt(message: ampq.ConsumeMessage) {
        if (!this.isConnected()) await this.connection.connect();
        this.connection.channel!.ack(message);
        logger.debug(
            `message ${message.fields.consumerTag}/${message.fields.deliveryTag} acknowledged`,
            { id: 'RabbitMQMessageReceiver/acknowledgeMessageReceipt' },
        );
    }
}

File: service-faces/src/modules/rabbitmq.py
import json

from datetime import datetime, timezone
from typing import Any, Callable, Generic, Optional, TypeVar, List, Awaitable
import asyncio

import aio_pika
from aio_pika.abc import (
    AbstractRobustConnection,
    AbstractRobustChannel,
)
from pydantic import BaseModel, ConfigDict, Field

from configs.config import config
from modules.logger import get_logger

T = TypeVar("T")

conn_info = config["rabbitmq"]["connection_settings"]
origin_queue_name = config["rabbitmq"]["service_queue_name"]


class RabbitMqMessage(BaseModel, Generic[T]):
    from_: str = Field(alias="from")
    to: str
    time: str
    jobId: str
    errors: List[str]
    filepath: str
    md5: str
    message: T

    model_config = ConfigDict(populate_by_name=True)


class RabbitMqConnection:
    def __init__(self, queue_name: str, durable: bool = True):
        self.queue_name = queue_name
        self.durable = durable
        self.connection: Optional[AbstractRobustConnection] = None
        self.channel: Optional[AbstractRobustChannel] = None

    async def connect(self):
        logger = get_logger("RabbitMqConnection/connect")
        if self.connection and self.channel:
            return
        logger.info(
            f"Connecting to RabbitMQ at {conn_info['host']}:{conn_info['port']} for queue {self.queue_name}"
        )
        self.connection = await aio_pika.connect_robust(**conn_info)
        self.channel = await self.connection.channel()
        await self.channel.declare_queue(self.queue_name, durable=self.durable)
        logger.info("RabbitMQ Connection Established")

    def is_connected(self) -> bool:
        return self.connection is not None and self.channel is not None

    async def close(self):
        logger = get_logger("RabbitMqConnection/close")
        logger.info(
            f"Closing RabbitMQ connection to {conn_info['host']}:{conn_info['port']} queue: {self.queue_name}"
        )
        if self.channel:
            await self.channel.close()
        if self.connection:
            await self.connection.close()
        self.channel = None
        self.connection = None


class RabbitMqConnectionManager:
    def __init__(self, queue_name: str, durable: bool = True):
        self.queue_name = queue_name
        self.connection = RabbitMqConnection(queue_name, durable)

    def get_queue_name(self) -> str:
        return self.queue_name

    def is_connected(self) -> bool:
        return self.connection.is_connected()

    async def close(self):
        await self.connection.close()


class RabbitMqMessageSender(RabbitMqConnectionManager):
    async def send_json_message(
        self,
        queue_name: str,
        message: T,
        filepath: str,
        md5: str,
        job_id: str = "",
        corr_id: str = "",
        jwe_token: str = "",
        errors: List[str] = [],
        persistent: bool = True,
    ):
        if not self.connection.is_connected():
            await self.connection.connect()

        message_to_send = RabbitMqMessage[T](
            from_=origin_queue_name,
            to=queue_name,
            filepath=filepath,
            md5=md5,
            jobId=job_id,
            time=datetime.now(timezone.utc).isoformat(),
            errors=errors,
            message=message,
        )
        json_bytes = json.dumps(message_to_send.model_dump(by_alias=True)).encode(
            "utf-8"
        )

        await self.connection.channel.default_exchange.publish(
            aio_pika.Message(
                body=json_bytes,
                delivery_mode=aio_pika.DeliveryMode.PERSISTENT
                if persistent
                else aio_pika.DeliveryMode.NOT_PERSISTENT,
                headers={
                    "x-correlation-id": corr_id,
                    "authorization": f"Bearer {jwe_token}",
                },
            ),
            routing_key=queue_name,
        )


class RabbitMqMessageReceiver(RabbitMqConnectionManager):
    def __init__(
        self, queue_name: str, durable: bool = True, auto_acknowledge: bool = False
    ):
        super().__init__(queue_name, durable)
        self.auto_acknowledge = auto_acknowledge

    async def get_messages_on_queue(
        self,
        callback: Callable[
            [RabbitMqMessage[Any], aio_pika.IncomingMessage], Awaitable[None]
        ],
    ):
        if not self.is_connected():
            await self.connection.connect()

        logger = get_logger("RabbitMqMessageReceiver/get_messages_on_queue")

        queue = await self.connection.channel.declare_queue(
            self.queue_name, durable=self.connection.durable
        )

        # Setup a future to allow breaking after 1 message in tests
        loop = asyncio.get_event_loop()
        message_handled = loop.create_future()

        async def _consumer(message: aio_pika.IncomingMessage):
            logger.info(f"Message received on queue: {self.queue_name}")
            try:
                raw = message.body.decode("utf-8")
                parsed_message = RabbitMqMessage[Any].model_validate_json(raw)

                await callback(parsed_message, message)

                if self.auto_acknowledge:
                    await message.ack()

                if not message_handled.done():
                    message_handled.set_result(True)

            except Exception as e:
                logger.error(f"Error processing message: {e}")
                await message.ack()

        await queue.consume(_consumer, no_ack=self.auto_acknowledge)

        try:
            await asyncio.wait_for(message_handled, timeout=3)
        except asyncio.TimeoutError:
            logger.warning("Timeout waiting for message")
